File: app/src/main/java/kr/djspi/pipe01/fragment/SurveyDialog.java
package kr.djspi.pipe01.fragment;

import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.DialogFragment;
import android.text.InputFilter;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.RadioButton;
import android.widget.RadioGroup;
import android.widget.TextView;

import com.helloelliote.filter.DecimalFilter;
import com.helloelliote.geolocation.GeoTrans;

import kr.djspi.pipe01.R;

import static java.util.Objects.requireNonNull;
import static kr.djspi.pipe01.Const.RESULT_FAIL;
import static kr.djspi.pipe01.Const.RESULT_PASS;
import static kr.djspi.pipe01.Const.TAG_SURVEY;

public class SurveyDialog extends DialogFragment implements OnClickListener {

    private static final String TAG = SurveyDialog.class.getSimpleName();
    private static final double INPUT_LIMIT = 999999.9999; // 최대 입력값
    public static GeoTrans.Coordinate originPoint;
    private final InputFilter[] FILTER_SURVEY = {new DecimalFilter(10, 4)};
    private int selectIndex = -1;
    private String dialogTitle;
    private RadioGroup radioGroup; // 원점 선택 라디오 버튼 그룹
    private TextInputLayout inputLayout_x, inputLayout_y;
    private TextInputEditText input_x, input_y;
    private OnSelectListener listener;

    public SurveyDialog() {
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnSelectListener) {
            listener = (OnSelectListener) context;
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        dialogTitle = getString(R.string.popup_title_survey);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_location_survey, container, false);

        TextView titleView = view.findViewById(R.id.popup_title);
        titleView.setText(dialogTitle);

        radioGroup = view.findViewById(R.id.nmap_radiogroup);
        radioGroup.setOnCheckedChangeListener((group, checkedId) -> {
            RadioButton checkedRadioButton = radioGroup.findViewById(checkedId);
            selectIndex = radioGroup.indexOfChild(checkedRadioButton);
            switch (checkedId) {
                case R.id.nmap_radio_central:
                    originPoint = GeoTrans.Coordinate.GRS80_MIDDLE_WITH_JEJUDO;
                    break;
                case R.id.nmap_radio_east:
                    originPoint = GeoTrans.Coordinate.GRS80_EAST;
                    break;
                case R.id.nmap_radio_eastsea:
                    originPoint = GeoTrans.Coordinate.GRS80_EASTSEA;
                    break;
                case R.id.nmap_radio_west:
                    originPoint = GeoTrans.Coordinate.GRS80_WEST;
                    break;
                default:
                    break;
            }
        });
        inputLayout_x = view.findViewById(R.id.lay_coordinate_x);
        inputLayout_y = view.findViewById(R.id.lay_coordinate_y);
        input_x = inputLayout_x.findViewById(R.id.input_coordinate_x);
        input_y = inputLayout_y.findViewById(R.id.input_coordinate_y);
        input_x.setFilters(FILTER_SURVEY);
        input_y.setFilters(FILTER_SURVEY);

        view.findViewById(R.id.btn_dismiss).setOnClickListener(this);
        view.findViewById(R.id.btn_ok).setOnClickListener(this);

        return view;
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.btn_ok:
                if (isInputValid(input_x, input_y)) {
                    assert input_x.getText() != null;
                    assert input_y.getText() != null;
                    listener.onSelect(TAG_SURVEY, RESULT_PASS,
                            input_x.getText().toString(),
                            input_y.getText().toString());
                    dismissAllowingStateLoss();
                } else selectIndex = -1;
                break;
            case R.id.btn_dismiss:
                listener.onSelect(TAG_SURVEY, RESULT_FAIL, (String) null);
                dismissAllowingStateLoss();
                break;
            default:
                break;
        }
    }

    /**
     * 사용자의 입력값이 유효한 값인지 검사: Null 체크, 유효범위 체크
     *
     * @param input_x
     * @param input_y
     * @return boolean isX, boolean isY 입력값 X AND Y 가 유효하면 true 리턴
     */
    private boolean isInputValid(TextInputEditText input_x, TextInputEditText input_y) {
        boolean isX = false;
        boolean isY = false;

        try {
            double value_x = Double.valueOf(requireNonNull(input_x.getText()).toString());
            if (value_x > INPUT_LIMIT) {
                inputLayout_x.setError(getString(R.string.map_coord_error));
            } else {
                isX = true;
                inputLayout_x.setError(null);
                input_x.clearFocus();
            }
        } catch (NullPointerException | NumberFormatException e) {
            inputLayout_x.setError(getString(R.string.map_input_error));
        }

        try {
            double value_y = Double.valueOf(requireNonNull(input_y.getText()).toString());
            if (value_y > INPUT_LIMIT) {
                inputLayout_y.setError(getString(R.string.map_coord_error));
            } else {
                isY = true;
                inputLayout_y.setError(null);
                input_y.clearFocus();
            }
        } catch (NullPointerException | NumberFormatException e) {
            inputLayout_y.setError(getString(R.string.map_input_error));
        }
        return (isX && isY);
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        selectIndex = -1;
        super.onDismiss(dialog);
    }

    @Override
    public void onDetach() {
        super.onDetach();
        listener = null;
    }
}

File: app/src/main/java/kr/djspi/pipe01/fragment/SurveyDialog.kt
package kr.djspi.pipe01.fragment

import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RadioButton
import android.widget.RadioGroup
import android.widget.TextView
import androidx.fragment.app.DialogFragment
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import kr.djspi.pipe01.Const.RESULT_FAIL
import kr.djspi.pipe01.Const.RESULT_PASS
import kr.djspi.pipe01.Const.TAG_SURVEY
import kr.djspi.pipe01.R
import kr.djspi.pipe01.geolocation.GeoTrans
import kr.djspi.pipe01.geolocation.GeoTrans.Coordinate.*
import kr.djspi.pipe01.util.DecimalFilter

class SurveyDialog : DialogFragment(), View.OnClickListener {

    private var dialogTitle: String? = null
    private var selectIndex: Int = -1
    private var checkeId: Int = -1
    private lateinit var inputX: TextInputEditText
    private lateinit var inputY: TextInputEditText
    private lateinit var layX: TextInputLayout
    private lateinit var layY: TextInputLayout
    private lateinit var listener: OnSelectListener

    override fun onAttach(context: Context) {
        super.onAttach(context)
        if (context is OnSelectListener) {
            listener = context
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        dialogTitle = getString(R.string.popup_title_survey)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_location_survey, container, false)
        val popupTitle = view.findViewById<TextView>(R.id.popup_title)
        popupTitle.text = dialogTitle
        val nmapRadiogroup = view.findViewById<RadioGroup>(R.id.nmap_radioGroup)
        nmapRadiogroup.setOnCheckedChangeListener { group, checkedId ->
            val checkedRadioButton = group.findViewById<RadioButton>(checkedId)
            checkeId = checkedId
            selectIndex = group.indexOfChild(checkedRadioButton)
            when (checkedId) {
                R.id.nmap_radio_central -> originPoint = GRS80_MIDDLE_WITH_JEJUDO
                R.id.nmap_radio_east -> originPoint = GRS80_EAST
                R.id.nmap_radio_eastsea -> originPoint = GRS80_EASTSEA
                R.id.nmap_radio_west -> originPoint = GRS80_WEST
            }
        }
        inputX = view.findViewById(R.id.input_x)
        inputY = view.findViewById(R.id.input_y)
        inputX.filters = arrayOf(DecimalFilter(10, 4))
        inputY.filters = arrayOf(DecimalFilter(10, 4))
        layX = view.findViewById(R.id.lay_x)
        layY = view.findViewById(R.id.lay_y)
        arrayOf<View>(
            view.findViewById(R.id.button_dismiss),
            view.findViewById(R.id.btn_ok)
        ).forEach {
            it.setOnClickListener(this)
        }
        return view
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.btn_ok -> {
                if (checkeId == -1) {
                    layX.error = "원점을 선택해주세요."
                    layY.error = "원점을 선택해주세요."
                    return
                }
                if (isInputValid()) {
                    listener.onSelect(
                        TAG_SURVEY, RESULT_PASS,
                        inputX.text.toString(),
                        inputY.text.toString()
                    )
                    dismiss()
                } else selectIndex = -1
            }
            R.id.button_dismiss -> {
                listener.onSelect(TAG_SURVEY, RESULT_FAIL, null)
                dismissAllowingStateLoss()
            }
        }
    }

    /**
     * 사용자의 입력값이 유효한 값인지 검사: Null 체크, 유효범위 체크
     *
     * @return boolean isX, boolean isY 입력값 X AND Y 가 유효하면 true 리턴
     */
    private fun isInputValid(): Boolean {
        var isX = false
        var isY = false

        try {
            if (inputX.text.toString().toDouble() > 999999.9999) {
                layX.error = getString(R.string.map_coord_error)
            } else {
                isX = true
                layX.error = null
                inputX.clearFocus()
            }
        } catch (e: NullPointerException) {
            layX.error = getString(R.string.map_input_error)
        } catch (e: NumberFormatException) {
            layX.error = getString(R.string.map_input_error)
        }

        try {
            if (inputY.text.toString().toDouble() > 999999.9999) {
                layY.error = getString(R.string.map_coord_error)
            } else {
                isY = true
                layY.error = null
                inputY.clearFocus()
            }
        } catch (e: NullPointerException) {
            layY.error = getString(R.string.map_input_error)
        } catch (e: NumberFormatException) {
            layY.error = getString(R.string.map_input_error)
        }
        return isX && isY
    }

    override fun onDismiss(dialog: DialogInterface) {
        selectIndex = -1
        checkeId = -1
        super.onDismiss(dialog)
    }

    companion object {
        lateinit var originPoint: GeoTrans.Coordinate
    }
}
